### PR TITLE
[autoscaler][Kubernetes] Operator subprocess error reporting, configuration fixes

### LINKED ISSUE #13589: [kubernetes] ray crashes when using the k8s operator and setting redis configurations
I'm using:
```
root@continual-ray-head-phv5r:/app/continual/services/ray_services# ray --version
ray, version 2.0.0.dev0
```
and
```
kubectl version
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.0", GitCommit:"af46c47ce925f4c4ad5cc8d1fca46c7b77d13b38", GitTreeState:"clean", BuildDate:"2020-12-08T17:59:43Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-07T21:12:17Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"linux/amd64"}
```

### What is the problem?

when testing out the example cluster yamls for k8s using the new operator if I add a redis configuration to the headStartRayCommands it crashes.  If I add the password like:

```
headStartRayCommands:
    - ray stop
    - ulimit -n 65536; ray start --head --port=6379 --object-manager-port=8076  --dashboard-host 0.0.0.0  --redis-password='password'
```

I get:
```
2021-01-21 00:17:45,153 INFO services.py:1172 -- View the Ray dashboard at http://172.17.0.8:8265
2021-01-21 00:17:44,529 INFO scripts.py:537 -- Local node IP: 0.0.0.0
2021-01-21 00:17:45,179 SUCC scripts.py:565 -- --------------------
2021-01-21 00:17:45,179 SUCC scripts.py:566 -- Ray runtime started.
2021-01-21 00:17:45,179 SUCC scripts.py:567 -- --------------------
2021-01-21 00:17:45,180 INFO scripts.py:569 -- Next steps
2021-01-21 00:17:45,180 INFO scripts.py:570 -- To connect to this Ray runtime from another node, run
2021-01-21 00:17:45,180 INFO scripts.py:574 --   ray start --address='0.0.0.0:6379' --redis-password='password'
2021-01-21 00:17:45,180 INFO scripts.py:579 -- Alternatively, use the following Python code:
2021-01-21 00:17:45,180 INFO scripts.py:582 -- import ray
2021-01-21 00:17:45,180 INFO scripts.py:583 -- ray.init(address='auto', _redis_password='password')
2021-01-21 00:17:45,180 INFO scripts.py:591 -- If connection fails, check your firewall settings and network configuration.
2021-01-21 00:17:45,180 INFO scripts.py:596 -- To terminate the Ray runtime, run
2021-01-21 00:17:45,180 INFO scripts.py:597 --   ray stop
[2021-01-20 16:17:45,471 C 3758 3758] redis_context.cc:361:  Check failed: _s.ok() Bad status: RedisError: 
*** StackTrace Information ***
    @     0x7f580b402bc5  google::GetStackTraceToString()
    @     0x7f580b3d2a1e  ray::GetCallTrace()
    @     0x7f580b3f8104  ray::RayLog::~RayLog()
    @     0x7f580b16fbda  ray::gcs::RedisContext::Connect()
    @     0x7f580b167720  ray::gcs::RedisClient::Connect()
    @     0x7f580b16855b  ray::gcs::RedisClient::Connect()
    @     0x7f580b0fd852  ray::gcs::ServiceBasedGcsClient::Connect()
    @     0x7f580afe7a8d  ray::gcs::GlobalStateAccessor::Connect()
    @     0x7f580af1f5ab  __pyx_pw_3ray_7_raylet_19GlobalStateAccessor_3connect()
    @     0x56310e82ab71  _PyMethodDef_RawFastCallKeywords
    @     0x56310e831aef  _PyMethodDescr_FastCallKeywords
    @     0x56310e89637c  _PyEval_EvalFrameDefault
    @     0x56310e7da2b9  _PyEval_EvalCodeWithName
    @     0x56310e7db610  _PyFunction_FastCallDict
    @     0x56310e7f9b93  _PyObject_Call_Prepend
    @     0x56310e8310aa  slot_tp_init
    @     0x56310e831ca8  _PyObject_FastCallKeywords
    @     0x56310e896d78  _PyEval_EvalFrameDefault
    @     0x56310e82a20b  _PyFunction_FastCallKeywords
    @     0x56310e891e70  _PyEval_EvalFrameDefault
    @     0x56310e7db31b  _PyFunction_FastCallDict
    @     0x56310e7f9b93  _PyObject_Call_Prepend
    @     0x56310e7ec95e  PyObject_Call
    @     0x56310e89351a  _PyEval_EvalFrameDefault
    @     0x56310e82a20b  _PyFunction_FastCallKeywords
    @     0x56310e891e70  _PyEval_EvalFrameDefault
    @     0x56310e82a20b  _PyFunction_FastCallKeywords
    @     0x56310e891e70  _PyEval_EvalFrameDefault
    @     0x56310e82a20b  _PyFunction_FastCallKeywords
    @     0x56310e891e70  _PyEval_EvalFrameDefault
    @     0x56310e7db31b  _PyFunction_FastCallDict
    @     0x56310e7f9b93  _PyObject_Call_Prepend
```

### Reproduction (REQUIRED)

To reproduce, run the example_cluster.yaml and the process found at: https://docs.ray.io/en/master/cluster/k8s-operator.html

modify the headStartRayCommands as above and watch the logs in the ray-operator-pod.

### LINKED ISSUE #14699: [kubernetes] Allow exposing custom ports with Kubernetes Operator
The [Ray Kubernetes Operator](https://docs.ray.io/en/releases-1.2.0/cluster/k8s-operator.html) makes it easy to spin up a Ray cluster on Kubernetes, but unlike the cluster config, it doesn't provide a mechanism to expose custom ports.

Using a normal Ray config for Kubernetes, we could do something like this to add port `8000` for Ray Serve:

```
services:
      - apiVersion: v1
        kind: Service
        metadata:
            name: local-cluster-ray-head
        spec:
            selector:
                component: local-cluster-ray-head
            ports:
                - name: client
                  protocol: TCP
                  port: 10001
                  targetPort: 10001
                - name: dashboard
                  protocol: TCP
                  port: 8265
                  targetPort: 8265
                - name: serve
                  protocol: TCP
                  port: 8000
                  targetPort: 8000
```

However, the k8s operator abstracts the service definition, so we cannot add the 8000 port.

Instead, we need to create an entirely different service, which works, but is not ideal:

```
apiVersion: v1
kind: Service
metadata:
  name: local-cluster-ray-head-serve
  labels:
    app: local-cluster-ray-head-serve
  namespace: ray
spec:
  type: ClusterIP
  ports:
    - name: "8000"
      port: 8000
      targetPort: 8000
  selector:
    component: local-cluster-ray-head
```

cc @richardliaw

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Why are these changes needed?

Errors in the operator's monitor subprocesses are not handled at all currently.
This PR improves the situation slightly by reporting a failed monitor subprocesses in the status field of the corresponding RayCluster API object.

This PR does the following:

- If a monitor subprocess errors, it signals its error status by putting data into a `multiprocessing.Queue`.
  - A thread in the main process gets from the queue and updates the custom resource's status accordingly. 
- Allow monitor subprocesses to end normally instead of terminating them, using a `multiprocessing.Event`. (Heeding the warning in the multiprocessing [docs](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Process.terminate).)
- Adds a test that the error reporting is working: kill a Ray cluster's head pod, wait for the monitor to error, read the Error status from the CR. 

This PR also

- Raises a value error if a user tries to supply a redis password in ray start commands
- Reads the port from head ray start commands to pass to the monitor.

This PR also

- Adds a field `headServicePorts` to the CRD, allowing users to override the default ports. 

## Related issue number

Closes https://github.com/ray-project/ray/issues/13589
Closes https://github.com/ray-project/ray/issues/14699

Build directly on top of https://github.com/ray-project/ray/pull/15469 , to be reviewed after that PR.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by DmitriGekhtman** [comment]
Just going to add the remaining non-Helm-related planned Kubernetes improvements to this PR.

I've added logic that
- Raises a value error if a user tries to supply a redis password in the start commands
- Reads the port from the head's Ray start command, to pass to the monitor in the operator pod
(not going to ) 

Remains to be done in this PR:
configurability of ports for the head service

**Comment by DmitriGekhtman** [comment]
Head service port configurability added, done adding features to this PR.

**Comment by edoakes** [comment]
@DmitriGekhtman you have a bunch of merge conflicts from the other recent changes.

@yiranwang52 could you review this please?

File: python/ray/ray_operator/operator.py [hunk 1]
```diff
@@ -154,6 +159,14 @@ def run_event_loop():
...
         handle_event(event_type, cluster_cr, cluster_name, cluster_namespace)
 
+    # Run control loop.
+    for event in raycluster_cr_stream:
+        cluster_cr = event["object"]
+        cluster_name = cluster_cr["metadata"]["name"]
+        cluster_namespace = cluster_cr["metadata"]["namespace"]
+        event_type = event["type"]
+        handle_event(event_type, cluster_cr, cluster_name, cluster_namespace)
+
```
Comment (DmitriGekhtman) [inline python/ray/ray_operator/operator.py:169]: woops somehow managed to duplicate this.

**Comment by DmitriGekhtman** [comment]
cc @ijrsvt @ckw017 @AmeerHajAli, if you'd like to check this out too

**Comment by DmitriGekhtman** [comment]
(Master merged properly, good to review.)

File: python/ray/autoscaler/_private/monitor.py [hunk 2]
```diff
@@ -203,6 +210,10 @@ def _run(self):
             # round of messages.
             time.sleep(AUTOSCALER_UPDATE_INTERVAL_S)
 
+    def _stop_message_recieved(self):
+        return (self.stop_receiver_connection
+                and self.stop_receiver_connection.poll())
+
```
Comment (ijrsvt) [inline python/ray/autoscaler/_private/monitor.py:216]: Where is this used?

File: python/ray/ray_operator/operator.py [hunk 3]
```diff
@@ -205,7 +210,17 @@ def cluster_action(event_type: str, cluster_cr: Dict[str, Any],
...
 
+def status_handling_loop():
+    while True:
+        cluster_name, cluster_namespace, status = cluster_status_q.get()
+        operator_utils.set_status(cluster_name, cluster_namespace, status)
+
+
+status_handler = threading.Thread(target=status_handling_loop, daemon=True)
+
+
```
Comment (ijrsvt) [inline python/ray/ray_operator/operator.py:221]: Can you define this in main? (or at least not in the file's namespace).

File: python/ray/autoscaler/_private/monitor.py (hunk 2 above)
Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/monitor.py:216]: Thanks -- Had originally implemented the stop signaling with a mp.Pipe -- forgot to remove.

**Comment by DmitriGekhtman** [comment]
The next step after this PR is to restart after the error state is registered.
https://github.com/ray-project/ray/issues/15608
https://github.com/ray-project/ray/issues/14634

**Review by yiranwang52 — APPROVED** [review]

File: python/ray/ray_operator/operator.py [hunk 4]
```diff
@@ -55,29 +60,30 @@ def set_generation(self, generation: int) -> None:
...
         # First stop the subprocess if it's alive
         self.clean_up_subprocess()
         # Reinstantiate process with f as target and start.
         self.subprocess = mp.Process(
             name=self.subprocess_name, target=f, daemon=True)
         self.subprocess.start()
-        if wait_to_finish:
-            self.subprocess.join()
 
     def clean_up_subprocess(self):
```
Comment (yiranwang52) [inline python/ray/ray_operator/operator.py:71]: is this triggered if the CRD for a cluster got deleted, but some other clusters are still running?

Comment (DmitriGekhtman) [reply python/ray/ray_operator/operator.py:71]: Yeah, a CRD `"DELETED"` event triggers clean-up [here](https://github.com/ray-project/ray/blob/9584593ac6d443812611f76a5408b26d11cdf2b3/python/ray/ray_operator/operator.py#L210).

**Comment by DmitriGekhtman** [comment]
Will wait for CI to run again, running e2e tests once more.

**Comment by DmitriGekhtman** [comment]
e2e tests are fine...waiting for ci

**Comment by DmitriGekhtman** [comment]
@edoakes 
This is good to merge. I'm pretty sure mac wheel build failure does not have to do with this PR.

**Pull request merged**